File: src/core/validators/DomainEntitySubclass/DomainEntitySubclassIdentityRenameMustNotExistForMultiPropertyIdentity.js
// @flow
import { errorRuleBase } from '../ValidationRuleBase';
import { includeRuleBase } from '../ValidationRuleRepository';
import { MetaEdGrammar } from '../../../grammar/gen/MetaEdGrammar';
import { getProperty } from '../ValidationHelper';
import type SymbolTable from '../SymbolTable';
import SymbolTableEntityType from '../SymbolTableEntityType';

function valid(ruleContext: any, symbolTable: SymbolTable) : boolean {
  if (!ruleContext.property().some(x => getProperty(x).propertyComponents().propertyAnnotation().identityRename() != null)) return true;

  const baseEntityContext = symbolTable.get(SymbolTableEntityType.domainEntity(), ruleContext.baseName().getText());
  if (baseEntityContext == null) return true;

  return Array.from(baseEntityContext.propertySymbolTable.values())
    .filter(x => x.propertyComponents().propertyAnnotation().identity() != null).length <= 1;
}

// eslint-disable-next-line no-unused-vars
function failureMessage(ruleContext: any, symbolTable: SymbolTable) : string {
  const baseIdentifier = ruleContext.baseName().getText();
  return `Domain Entity '${ruleContext.entityName().getText()}' based on '${baseIdentifier}' is invalid for identity rename because parent entity '${baseIdentifier}' has more than one identity property.`;
}

const validationRule = errorRuleBase(valid, failureMessage);
// eslint-disable-next-line import/prefer-default-export
export const includeRule = includeRuleBase(MetaEdGrammar.RULE_domainEntitySubclass, validationRule);

File: src/core/validators/DomainEntitySubclass/DomainEntitySubclassIdentityRenameMustNotExistForMultiPropertyIdentity.cs
using System.Linq;
using MetaEd.Grammar.Antlr;

namespace MetaEd.Core.Validator.DomainEntitySubclass
{
    public class DomainEntitySubclassIdentityRenameMustNotExistForMultiPropertyIdentity : ValidationRuleBase<MetaEdGrammar.DomainEntitySubclassContext>
    {
        private readonly ISymbolTable _symbolTable;

        public DomainEntitySubclassIdentityRenameMustNotExistForMultiPropertyIdentity(ISymbolTable symbolTable)
        {
            _symbolTable = symbolTable;
        }

        public override bool IsValid(MetaEdGrammar.DomainEntitySubclassContext context)
        {
            var anyIdentityRenames =
                context.property()
                    .Any(x => x.GetProperty().propertyComponents().propertyAnnotation().identityRename() != null);
            if (!anyIdentityRenames)
                return true;

            var baseIdentifier = context.baseName().GetText();

            var baseSymbolTable = _symbolTable.Get(SymbolTableEntityType.DomainEntityEntityType(), baseIdentifier);
            if (baseSymbolTable == null)
                return true; // this will cause a validation failure somewhere else

            // Zero case is ok here because the Identity Property Reference check will handle that
            return
                baseSymbolTable.PropertySymbolTable.Values()
                    .Count(v => v.propertyComponents().propertyAnnotation().identity() != null) <= 1;
        }

        public override string GetFailureMessage(MetaEdGrammar.DomainEntitySubclassContext context)
        {
            var identifier = context.entityName().GetText();
            var baseIdentifier = context.baseName().GetText();
            
            return string.Format("Domain Entity '{0}' based on '{1}' is invalid for identity rename because parent entity '{1}' has more than one identity property.", identifier, baseIdentifier);
        }
    }
}
